### PR TITLE
Delay authentication plugin for keycloak

### PR DESCRIPTION
**Pull request by jjaferson**
Full workable version of the plugin

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -21,3 +21,10 @@
 
 # virtual machine crash logs, see http://www.java.com/en/download/help/error_hotspot.xml
 hs_err_pid*
+<<<<<<< HEAD
+
+# MacOS files
+.DS_Store
+=======
+/target/
+>>>>>>> add maven web project structure
```
Comment (matskiv) [inline .gitignore]: Unresolved conflict^

File: src/main/java/org/keycloak/plugin/rhmi/DelayAuthenticationAuthenticatorFactory.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+	
+    private static final DelayAuthentication SINGLETON = new DelayAuthentication();
+
+    public Authenticator create(KeycloakSession session) {
+        return SINGLETON;
+	}
+
+	public void init(Scope config) {
+		// TODO Auto-generated method stub
```
Comment (matskiv) [inline src/main/java/org/keycloak/plugin/rhmi/DelayAuthenticationAuthenticatorFactory.java]: We should remove all `// TODO` comments

File: docker-compose.yaml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  #   image: mysql:5.7
+  #   ports:
+  #     - "3306:3306"
+  #   environment:
+  #     - MYSQL_DATABASE=keycloak
+  #     - MYSQL_USER=keycloak
+  #     - MYSQL_PASSWORD=password
+  #     - MYSQL_ROOT_PASSWORD=root_password
+  #   volumes:
+  #     - data:/var/lib/mysql
```
Comment (matskiv) [inline docker-compose.yaml]: This^ can be removed?

**Comment by matskiv** [comment]
The `src/main/webapp/` files are unused, we can remove them, as well as related dependencies and configs.

File: .classpath [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+		<attributes>
+			<attribute name="maven.pomderived" value="true"/>
+		</attributes>
+	</classpathentry>
+	<classpathentry excluding="**" kind="src" output="target/classes" path="src/main/resources">
+		<attributes>
+			<attribute name="maven.pomderived" value="true"/>
+		</attributes>
+	</classpathentry>
+	<classpathentry kind="src" output="target/test-classes" path="src/test/java">
```
Comment (matskiv) [inline .classpath]: We don't have `src/test`

File: src/main/java/org/keycloak/plugin/rhmi/DelayAuthentication.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        context.setUser(federatedUser);
+
+        if (isUserCreated(context.getUser())) {
+        	redirectToAfterFirstBrokerLoginSuccess(context, brokerContext);
+		} else {
+			redirectToAction(context);
+		}
+	}
+	
+	private void redirectToAction(AuthenticationFlowContext context) {
```
Comment (matskiv) [inline src/main/java/org/keycloak/plugin/rhmi/DelayAuthentication.java]: I think this name isn't very precise. Maybe something like `showAccountProvisioningPage` ?

File: README.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
+# KeyCloak Delay Authentication Flow
+
+Plugin to delay authentication of users until they get created in Keycloak and added in 3scale by RHMI operator.
+
+# System Requirements
+
+You need to have Keycloak 3.2.0.Final running.
```
Comment (matskiv) [inline README.md]: This can't be accurate, Keycloak is on version 9.0.2 right now.

File: docker-compose.yaml [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+      - JDBC_PARAMS='connectTimeout=30'
+      - DB_VENDOR=mariadb
+      - DB_ADDR=mariadb
+      - DB_DATABASE=keycloak
+      - DB_USER=keycloak
+      - DB_PASSWORD=password
+      - JGROUPS_DISCOVERY_PROTOCOL=JDBC_PING
+      - JGROUPS_DISCOVERY_PROPERTIES=datasource_jndi_name=java:jboss/datasources/KeycloakDS,info_writer_sleep_time=500
+    volumes:
+      - ./plugins:/opt/jboss/keycloak/providers
```
Comment (matskiv) [inline docker-compose.yaml:19]: This didn't work with the `./target:/opt/jboss/keycloak/providers` 
to remove the need for copying the .jar file?

File: docker-compose.yaml [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+      - ./tmp:/tmp
```
Comment (matskiv) [inline docker-compose.yaml]: This should probably be deleted.

**Comment by matskiv** [comment]
I am not sure if the `.settings/*`, `.project` and `target/*` should be commited or added to `.gitignore` instead

File: .classpath (hunk 4 above)
Comment (jjaferson) [reply .classpath]: I do and I don't know it was not added to git. I'll remove it all as we are not adding unit testing

File: README.md (hunk 6 above)
Comment (jjaferson) [reply README.md]: Copied it from the [action-token-required-action](https://github.com/keycloak/keycloak-quickstarts/tree/latest/action-token-required-action), should I add 9.0.2 version there or leave without version?

File: docker-compose.yaml (hunk 7 above)
Comment (jjaferson) [reply docker-compose.yaml:19]: there are other files in `/target/`  and not sure if can cause conflicts. I've changed the path that maven outputs the `.jar` file to `./plugins/`.

```	    
<plugin>
  <groupId>org.apache.maven.plugins</groupId>
	      <artifactId>maven-jar-plugin</artifactId>
	      <version>2.3.1</version>
	      <configuration>
	        <outputDirectory>./plugins/</outputDirectory>
	      </configuration>
	    </plugin>```

**Comment by jjaferson** [comment]
> I am not sure if the `.settings/*`, `.project` and `target/*` should be commited or added to `.gitignore` instead

Good question, I'll have a look at other java repos

**Comment by jjaferson** [comment]
> > I am not sure if the `.settings/*`, `.project` and `target/*` should be commited or added to `.gitignore` instead
> 
> Good question, I'll have a look at other java repos

I ended up adding them to the `.gitignore`

**Comment by jjaferson** [comment]
@matskiv Thanks for reviewing I had a look through them and pushed the changes

File: README.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
+# KeyCloak Delay Authentication Flow
+
+Plugin to delay authentication of users until they get created in Keycloak and added in 3scale by RHMI operator.
+
+# System Requirements
+
+You need to have Keycloak 9.0.2 running.
```
Comment (matskiv) [inline README.md:7]: Now that I think of it, Keycloak is not a requirement because we use docker-compose...

**Review by matskiv — APPROVED** [review]
I tested the building process locally 👍 
LGTM, merging

**Pull request merged**